### PR TITLE
Add Matrix parsing

### PR DESCRIPTION
**Pull request by DrJosh9000**
Here's my take on it. It's maybe a bit heavy in named types ("`MatrixAdjustmentWith`", very Enterprise Software™), but this accounts for all the ways to write a matrix that I could find through the docs and playing around.

`MatrixScalars` basically exists to validate input. I had previously included the capability to sort `MatrixScalars`, but it isn't needed yet. (Or probably at all. Ordering ought to be stable throughout. I thought a lot about ~~The Matrix Canon~~ canonicalising matrices today.) `MatrixScalars` and `MatrixAdjustmentWith` limit values to string, bool, and int because those are the types that the backend accepts (which, coincidentally, will make it easier to produce a canonical encoding of `Matrix` for signing.)

### CODE REVIEW

**Comment by moskyb** [comment]
oh man, this is miles better than mine. i'll close it.

File: internal/pipeline/parser_test.go [hunk 1]
```diff
@@ -1488,3 +1488,438 @@ steps:
...
+    }
+  ]
+}`,
+		},
+	}
+
+	for _, test := range tests {
+		test := test
+		t.Run(test.desc, func(t *testing.T) {
+			//t.Parallel()
```
Comment (moskyb) [inline internal/pipeline/parser_test.go:1905]: should we parallelise these tests?

File: internal/pipeline/parser_test.go [hunk 2]
```diff
@@ -1488,3 +1488,438 @@ steps:
...
+			gotJSON, err := json.MarshalIndent(got, "", "  ")
+			if err != nil {
+				t.Fatalf(`json.MarshalIndent(got, "", "  ") error = %v`, err)
+			}
+			if diff := cmp.Diff(string(gotJSON), test.wantJSON); diff != "" {
+				t.Errorf("marshalled JSON diff (-got +want):\n%s", diff)
+			}
+		})
+	}
+}
```
Comment (moskyb) [inline internal/pipeline/parser_test.go:1925]: this file is getting really long - would it make sense to split it up a bit for easier navigation? happy for this to be done as part of a separate PR

File: internal/pipeline/step_command.go [hunk 3]
```diff
@@ -25,7 +25,7 @@ type CommandStep struct {
 	Plugins   Plugins           `yaml:"plugins,omitempty"`
 	Env       map[string]string `yaml:"env,omitempty"`
 	Signature *Signature        `yaml:"signature,omitempty"`
-	Matrix    any               `yaml:"matrix,omitempty"`
+	Matrix    *Matrix           `yaml:"matrix,omitempty"`
```
Comment (moskyb) [inline internal/pipeline/step_command.go:28]: 🥳

**Review by moskyb — COMMENTED** [review]
this is looking really good! i have a couple of comments and niggles, but otherwise it's looking 🔥

File: internal/pipeline/step_command_matrix.go [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+
+	_ interface {
+		json.Marshaler
+		selfInterpolater
+	} = (*MatrixAdjustment)(nil)
+
+	_ = []interface {
+		json.Marshaler
+		ordered.Unmarshaler
+		yaml.Marshaler
```
[the remark this answers is not in the document]
Comment (DrJosh9000) [reply internal/pipeline/step_command_matrix.go:27]: An adjustment always has `with` and sometimes `skip` or `soft_fail`. That's a struct that `ordered.Unmarshal` knows how to fill. `MatrixAdjustmentWith` can either be a map or a single value, and presently `ordered.Unmarshal` can't create "singleton maps". That wouldn't be hard to add though...

File: internal/pipeline/parser_test.go [hunk 5]
```diff
@@ -1488,3 +1488,438 @@ steps:
...
+			},
+			wantJSON: `{
+  "steps": [
+    {
+      "command": "echo {{matrix}}",
+      "matrix": {
+        "setup": [
+          "apple",
+          47,
+          true
```
[the remark this answers is not in the document]
Comment (DrJosh9000) [reply internal/pipeline/parser_test.go:1531]: That's possible, if there are no adjustments (which we also don't document in that case). Thinking...

File: internal/pipeline/parser_test.go (hunk 1 above)
Comment (DrJosh9000) [reply internal/pipeline/parser_test.go]: Done.

File: internal/pipeline/parser_test.go (hunk 2 above)
Comment (DrJosh9000) [reply internal/pipeline/parser_test.go]: I've moved the new tests, but the rest can remain for now.

File: internal/pipeline/parser_test.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (DrJosh9000) [reply internal/pipeline/parser_test.go]: Done (if the matrix is simple, then it becomes a list again).

**Review by moskyb — APPROVED** [review]
